### ISSUE TITLE
Zombie processes because of `SIGHeartbeatSender::unregister`

### ISSUE THREAD

**Issue by mrtus** [php-amqplib/php-amqplib#1204]
**Version(s) affected**: N/A
**PHP version**: N/A
**RabbitMQ version**: N/A

**Description**  
When using a heartbeat sender, `posix_kill` is used to kill the child process.
However this results in zombie processes that stay active because `pcntl_waitpid` is not called.

While maybe `SIGHeartbeatSender` is not meant to be ran in a PHP-FPM process, it sometimes has a use, we have FPM processes that can take some time, longer than the configured heartbeats, meaning the connection is always broken before a message is pushed on the connection

**How to reproduce**  
Run a PHP-FPM process that starts a heartbeat sender on every request.
Call `ps aux` and you'll see `php-fpx [defunct]` processes (aka zombie processes)

**Possible Solution**  
Call `pcntl_waitpid` after `posix_kill` is called;

https://github.com/php-amqplib/php-amqplib/blob/master/PhpAmqpLib/Connection/Heartbeat/SIGHeartbeatSender.php#L58

Let me know your thoughts on this topic!
Thanks

**Comment by ramunasd** [comment]
is method `unregister()` called before shutting down php-fpm process? This heartbeat sender does not ensure clean shutdown automatically.

**Comment by mrtus** [comment]
> is method `unregister()` called before shutting down php-fpm process? This heartbeat sender does not ensure clean shutdown automatically.

Correct! I forgot to mention. The `unregister` method is called in a `register_shutdown_function` and I can confirm this is being executed. `SIGHeartbeatSender::unregister` contains the `posix_kill`.

**Comment by ramunasd** [comment]
Can you confirm that `pcntl_waitpid()` inside shutdown function works as expected? I think it should be possible to use more general `pcntl_wait()` if cihld pid is not known.

**Comment by mrtus** [comment]
> Can you confirm that `pcntl_waitpid()` inside shutdown function works as expected? I think it should be possible to use more general `pcntl_wait()` if child pid is not known.

We are using `pcntl_waitpid($childId)` in production for our own, similar heartbeat sender but for PDO, and it behaves as expected as there are no longer any zombie processes after each `unregister`.

I also tested this on the AMQP heartbeat sender and I confirm that `pcntl_waitpid($childId)` also there resolves the problem of zombie processes after `unregister` is called.

**Comment by ramunasd** [comment]
Possible fix in PR https://github.com/php-amqplib/php-amqplib/pull/1205

**Comment by ramunasd** [comment]
fix released in version 3.7.3
